### PR TITLE
TR4 squishy block 1 + 2

### PR DESCRIPTION
**Pull request by Nemoel-Tomo**
Squishy block stops at a wall, raised floor, or lowered ceiling and then return. You can change the speed with the OCB value. OCB 0 = original behaviour

[squishyblock_new3.zip](https://github.com/MontyTRC89/TombEngine/files/14665740/squishyblock_new3.zip)

### CODE REVIEW

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+			return;
+
+		item.ItemFlags[0] = item.TriggerFlags;
+		auto forwardDir = EulerAngles(0, item.Pose.Orientation.y + item.ItemFlags[4], 0).ToDirection();		
+	
+		auto pointColl = GetCollision(item.Pose.Position.x, item.Pose.Position.y, item.Pose.Position.z, item.RoomNumber);
+
+		// DEBUG
+		auto bounds = GameBoundingBox(&item);
+		g_Renderer.AddDebugBox(bounds.ToBoundingOrientedBox(item.Pose), Vector4::One, RendererDebugPage::None, true);
```
Comment (Jakub768) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:62]: Is this needed? Is it for when player presses F10?

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+			if (!IsNextSectorValid(item, forwardDir, item.ItemFlags[0]))
+			{
+				if (item.ItemFlags[4] == ANGLE(180))
+				{
+					item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_COLLIDE_LEFT;
+					item.Animation.FrameNumber = GetAnimData(item).frameBase;
+					item.Animation.ActiveState = SQUISHY_BLOCK_STATE_COLLIDE_LEFT;
+					item.Animation.TargetState = SQUISHY_BLOCK_STATE_COLLIDE_LEFT;
+
+					if (&item.TouchBits)
```
Comment (Jakub768) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:78]: this if statement is not completed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+					}
+				}
+				else if (item.ItemFlags[4] == ANGLE(0))
+				{
+					item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_COLLIDE_RIGHT;
+					item.Animation.FrameNumber = GetAnimData(item).frameBase;
+					item.Animation.ActiveState = SQUISHY_BLOCK_STATE_COLLIDE_RIGHT;
+					item.Animation.TargetState = SQUISHY_BLOCK_STATE_COLLIDE_RIGHT;
+
+					if (&item.TouchBits)
```
Comment (Jakub768) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:91]: same as above

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			{
+				slopeAngle = ANGLE(0.0f);
+			}
+
+			/*short dirAngle = phd_atan(dir.z, dir.x);
+			short alignAngle = Geometry::GetShortestAngle(slopeAngle, dirAngle);
+
+			// Test if slope aspect is aligned with direction.
+			if (alignAngle != 0 && alignAngle != ANGLE(180.0f))
+				return false;*/
```
Comment (Jakub768) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:167]: if the commented bit is not needed then it can be deleted :)

**Comment by Nemoel-Tomo** [comment]
there is something wrong with the y axis collision. I will check that

File: TombEngine/Game/collision/collide_item.cpp [hunk 5]
```diff
@@ -662,19 +662,24 @@ bool ItemPushItem(ItemInfo* item0, ItemInfo* item1, CollisionInfo* coll, bool en
...
 		maxZ += coll->Setup.Radius;
+		minY -= coll->Setup.Radius;
+		maxY += coll->Setup.Radius;
 	}
 
 	// Big enemies.
 	if (abs(deltaPos.x) > BLOCK(4.5f) || abs(deltaPos.z) > BLOCK(4.5f) ||
 		relDeltaPos.x <= minX || relDeltaPos.x >= maxX ||
-		relDeltaPos.z <= minZ || relDeltaPos.z >= maxZ)
+		relDeltaPos.z <= minZ || relDeltaPos.z >= maxZ ||
```
Comment (Lwmte) [inline TombEngine/Game/collision/collide_item.cpp:681]: Can you give an overview of this change? I suspect this has something to do with pushing item on vertical (Y) axis?

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+		if (!item.TriggerFlags)
+		{
+			SetAnimation(item, SQUISHY_BLOCK_ANIM_ORIGINAL);
+			item.Animation.AnimNumber = Objects[item.ObjectNumber].animIndex + SQUISHY_BLOCK_ANIM_ORIGINAL;
+			item.Animation.FrameNumber = GetAnimData(item).frameBase;
+			item.Animation.ActiveState = SQUISHY_BLOCK_STATE_ORIGINAL;
+			item.Animation.TargetState = SQUISHY_BLOCK_STATE_ORIGINAL;
+		}
+
+			item.ItemFlags[0] = item.TriggerFlags;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:48]: Please reduce tabulation in this and forthcoming 3 lines by one tab.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+						item.HitPoints = NOT_TARGETABLE;
+					}
+
+					item.ItemFlags[4] = item.ItemFlags[4] + ANGLE(180.0f);
+					item.Pose.Orientation.y += ANGLE(item.ItemFlags[1]);
+				}				
+			}
+		}
+			if (LaraItem->HitPoints)
+			AnimateItem(&item);
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:130]: Please add 1 tab here.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+					item.ItemFlags[4] = item.ItemFlags[4] + ANGLE(180.0f);
+					item.Pose.Orientation.y += ANGLE(item.ItemFlags[1]);
+				}				
+			}
+		}
+			if (LaraItem->HitPoints)
+			AnimateItem(&item);		
+	}
+	
+	bool IsNextSectorValid(const ItemInfo& item, const Vector3& dir)
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:133]: I think this function can be useful outside of the squishy block code, can we move it to some more generic place?

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+			if (laraItem->HitPoints > 0)
+				ItemPushItem(&item, laraItem, coll, false, 1);			
+		}
+
+		if (ItemPushItem(&item, laraItem, coll, false, 1))
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				auto frame = item.Animation.FrameNumber - GetAnimData(item).frameBase;
+				if (!frame || frame == 33)
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:234]: Can we somehow dehardcode magic number 33 here? If 33 is the last frame of the animation, you can use `TestLastFrame` helper function (look for usage examples in existing code).

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+					LaraItem->HitPoints = 0;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:235]: Please use `DoDamage(&laraItem, INT_MAX)` here instead. Also, avoid referencing `LaraItem` global (note that `LaraItem` starts with a capital L here) in your code, it is "toxic" global we should eventually get rid of.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+	void FallingSquishyBlockCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TestBoundsCollide(&item, laraItem, coll->Setup.Radius) && TestCollision(&item, laraItem))
+		{
+			if (item.Animation.FrameNumber - GetAnimData(item).frameBase <= 8)
+			{
+				item.Animation.FrameNumber += 2;
+				laraItem->HitPoints = 0;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:254]: Please use `DoDamage(&laraItem, INT_MAX)` here.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				auto frame = item.Animation.FrameNumber - GetAnimData(item).frameBase;
+				if (!frame || frame == 33)
+					LaraItem->HitPoints = 0;
+			}
+			else if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_COLLIDE_RIGHT ||
+					item.Animation.ActiveState == SQUISHY_BLOCK_STATE_COLLIDE_LEFT)
+			{
+				LaraItem->HitPoints = 0;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:240]: Same as above.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+	void ControlFallingSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TriggerActive(&item))
+		{
+			if (item.ItemFlags[0] < 60)
+			{
+				SoundEffect(SFX_TR4_EARTHQUAKE_LOOP, &item.Pose);
+				Camera.bounce = (item.ItemFlags[0] - 92) >> 1;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:273]: Can we avoid using bitshift here? Meaning of this operation is not clear in this context.

File: TombEngine/Game/collision/collide_item.cpp (hunk 5 above)
Comment (Sezzary) [reply TombEngine/Game/collision/collide_item.cpp]: This prevents an object's bounding box from essentially having infinite height. The player was getting pushed by blocks even if they were passing overhead.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp (hunk 8 above)
Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: I put it into misc.cpp

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp (hunk 9 above)
Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: fixed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp (hunk 13 above)
Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: fixed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+		{
+			if (item.Animation.ActiveState != SQUISHY_BLOCK_STATE_BAKED_MOTION)
+				SetAnimation(item, SQUISHY_BLOCK_ANIM_BAKED_MOTION);
+		}
+		else
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_ANIM_BAKED_MOTION)
+				SetAnimation(item, SQUISHY_BLOCK_STATE_MOVE);
+
+			something0 = item.TriggerFlags;
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:83]: This variable name is totally non-descriptive, please give it a meaningful name.

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+		}
+
+		AnimateItem(&item);		
+	}
+
+	void ControlFallingSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		short& something0 = item.ItemFlags[0];
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:137]: As above

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+		if (TriggerActive(&item))
+		{
+			if (something0 < 60)
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:141]: Please try to demagic all the magic numbers here as much as possible.

Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: fixed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp (hunk 15 above)
Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: fixed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp (hunk 14 above)
Comment (Nemoel-Tomo) [reply TombEngine/Objects/TR4/Trap/SquishyBlock.cpp]: fixed

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+				}
+				else
+				{
+					float dist = Lerp(item.TriggerFlags / 4, item.TriggerFlags, velocity);
+					item.Pose.Position = Geometry::TranslatePoint(item.Pose.Position, forwardDir, dist);
+				}
+			}
+			else
+			{
+				if ((item.Animation.FrameNumber - GetAnimData(item).frameBase) == 19)
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:117]: Please demagic this

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+		{
+			if (velocity < MAX_FALLING_VELOCITY)
+			{
+				SoundEffect(SFX_TR4_EARTHQUAKE_LOOP, &item.Pose);
+				Camera.bounce = (velocity - 92) / 2;
+				velocity++;
+			}
+			else
+			{
+				if ((item.Animation.FrameNumber - GetAnimData(item).frameBase) == 8)
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:150]: Please demagic this

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+		}	
+	}
+
+	void CollideFallingSquishyBlock(short itemNumber, ItemInfo* playerItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TestBoundsCollide(&item, playerItem, coll->Setup.Radius) && TestCollision(&item, playerItem))
+		{
+			if ((item.Animation.FrameNumber - GetAnimData(item).frameBase) <= 8)
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:190]: Please demagic this

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+			{
+				item.Animation.FrameNumber += 2;
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:192]: And again, please demagic this

File: TombEngine/Objects/TR4/Trap/SquishyBlock.cpp [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+#include "Game/misc.h"
+#include "Game/Setup.h"
+#include "Sound/sound.h"
+#include "Specific/level.h"
+
+// NOTES:
+// item.ItemFlags[0]: use dynamic motion.
+// item.ItemFlags[1]: ??
+// item.ItemFlags[4]: heading angle.
+// TODO: Add support for 4 directions.
```
Comment (Stranger1992) [inline TombEngine/Objects/TR4/Trap/SquishyBlock.cpp:22]: Is this TODO being worked on at this time or is it being worked on after?

**Review by Stranger1992 — CHANGES_REQUESTED** [review]
I have left some comments on the code (as you'll see).

I have a problem with the `SQUISHY1` in that an OCB above 300 will not crush lara when she is next to a wall. I recommend having an entry to the console to say "OCB of SQUISHY_BLOCK1 (insert Lua name) is above maximum. Reverting to maximum speed" (or something to that effect). The problem is illustrated below:

https://github.com/MontyTRC89/TombEngine/assets/84292688/f3a797bf-4c9f-45e1-a68c-ef66bfc44107

File: TombEngine/Scripting/Internal/TEN/Objects/ObjectIDs.h [hunk 22]
```diff
@@ -1519,8 +1519,8 @@ static const std::unordered_map<std::string, GAME_OBJECT_ID> kObjIDs {
 	{ "SARCOPHAGUS", ID_SARCOPHAGUS },
 	{ "ENEMY_PIECE", ID_ENEMY_PIECE },
 	{ "EXPANDING_PLATFORM", ID_EXPANDING_PLATFORM },
-	{ "SQUISHY_BLOCK1", ID_SQUISHY_BLOCK1 },
-	{ "SQUISHY_BLOCK2", ID_SQUISHY_BLOCK2 },
+	{ "SQUISHY_BLOCK1", ID_SQUISHY_BLOCK_HORIZONTAL },
```
Comment (Lwmte) [inline TombEngine/Scripting/Internal/TEN/Objects/ObjectIDs.h:1522]: Please rename left side too

**Review by Stranger1992 — APPROVED** [review]

**Review by Lwmte — APPROVED** [review]

**Review by Stranger1992 — APPROVED** [review]
Final testing:

All previously flagged issues have been resolved and this PR can now be merged! Well done @Nemoel-Tomo for your hard work 👍🏻

**Pull request merged**